controller panic after update with resource limit spec

<!--
Pro-tip: You can leave this block commented, and it still works!

Select the appropriate areas for your issue:

/area API

Classify what kind of issue this is:

/kind bug

-->

## Expected Behavior

controller  not painc

## Actual Behavior
controller panic  and exit



## Steps to Reproduce the Problem

1. create one service with resources spec

```
apiVersion: serving.knative.dev/v1alpha1
kind: Service
metadata:
  name: helloworld2
spec:
  runLatest:
    configuration:
      revisionTemplate:
        spec:
          container:
            image: PRIVATEHUB/serving/helloworld-go:0.1
            env:
            - name: TARGET
              value: "test  Go Sample v2"
            resources:
              requests:
                memory: "64Mi"
                cpu: "250m"
              limits:
                memory: "128Mi"
                cpu: "600m"
```

2. change the resource limit and apply again
3. controller panic


## Additional Info

The bug introduce from https://github.com/knative/serving/issues/2099

file: src/github.com/knative/serving/pkg/reconciler/v1alpha1/service/service.go
```
logger.Infof("Reconciling configuration diff (-desired, +observed): %v", cmp.Diff(desiredConfig.Spec, config.Spec))
```
this diff miss quantityComparer


panic log
```
panic: cannot handle unexported field: {v1alpha1.ConfigurationSpec}.RevisionTemplate.Spec.Container.Resources.Limits["cpu"].i
consider using AllowUnexported or cmpopts.IgnoreUnexported [recovered]
        panic: cannot handle unexported field: {v1alpha1.ConfigurationSpec}.RevisionTemplate.Spec.Container.Resources.Limits["cpu"].i
consider using AllowUnexported or cmpopts.IgnoreUnexported

goroutine 380 [running]:
github.com/knative/serving/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x119b360, 0xc420fc2320)
        /opt/go/src/runtime/panic.go:502 +0x229
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.invalid.apply(0xc421038230, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/options.go:208 +0xf5
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).tryOptions(0xc421038230, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x15129a0, 0x1312d80, 0x0)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:307 +0x149
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareAny(0xc421038230, 0x1312d80, 0xc420712800, 0xb9, 0x1312d80, 0xc420712840, 0xb9)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:202 +0x347
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareStruct(0xc421038230, 0x130db00, 0xc420712800, 0x99, 0x130db00, 0xc420712840, 0x99, 0x15129a0, 0x130db00)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:503 +0x171
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareAny(0xc421038230, 0x130db00, 0xc420712800, 0x99, 0x130db00, 0xc420712840, 0x99)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:272 +0x1976
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareMap(0xc421038230, 0x12642c0, 0xc420446cf8, 0x95, 0x12642c0, 0xc420446fb8, 0x95, 0x15129a0, 0x12642c0)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:460 +0x61b
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareAny(0xc421038230, 0x12642c0, 0xc420446cf8, 0x95, 0x12642c0, 0xc420446fb8, 0x95)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:269 +0x1644
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareStruct(0xc421038230, 0x1289c00, 0xc420446cf8, 0x99, 0x1289c00, 0xc420446fb8, 0x99, 0x15129a0, 0x1289c00)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:503 +0x171
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareAny(0xc421038230, 0x1289c00, 0xc420446cf8, 0x99, 0x1289c00, 0xc420446fb8, 0x99)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:272 +0x1976
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareStruct(0xc421038230, 0x1360a80, 0xc420446c50, 0x99, 0x1360a80, 0xc420446f10, 0x99, 0x15129a0, 0x1360a80)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:503 +0x171
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareAny(0xc421038230, 0x1360a80, 0xc420446c50, 0x99, 0x1360a80, 0xc420446f10, 0x99)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:272 +0x1976
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareStruct(0xc421038230, 0x1332240, 0xc420446bf8, 0x99, 0x1332240, 0xc420446eb8, 0x99, 0x15129a0, 0x1332240)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:503 +0x171
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareAny(0xc421038230, 0x1332240, 0xc420446bf8, 0x99, 0x1332240, 0xc420446eb8, 0x99)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:272 +0x1976
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareStruct(0xc421038230, 0x1286c80, 0xc420446b10, 0x99, 0x1286c80, 0xc420446dd0, 0x99, 0x15129a0, 0x1286c80)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:503 +0x171
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareAny(0xc421038230, 0x1286c80, 0xc420446b10, 0x99, 0x1286c80, 0xc420446dd0, 0x99)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:272 +0x1976
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareStruct(0xc421038230, 0x1292d20, 0xc420446b00, 0x99, 0x1292d20, 0xc420446dc0, 0x99, 0x15129a0, 0x1292d20)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:503 +0x171
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.(*state).compareAny(0xc421038230, 0x1292d20, 0xc420446b00, 0x99, 0x1292d20, 0xc420446dc0, 0x99)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:272 +0x1976
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.Equal(0x1292d20, 0xc420446b00, 0x1292d20, 0xc420446dc0, 0xc4206521e0, 0x2, 0x2, 0xc4210ff2a0)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:86 +0xf1
github.com/knative/serving/vendor/github.com/google/go-cmp/cmp.Diff(0x1292d20, 0xc420446b00, 0x1292d20, 0xc420446dc0, 0x0, 0x0, 0x0, 0xc420652100, 0xc420652100)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/google/go-cmp/cmp/compare.go:100 +0x133
github.com/knative/serving/pkg/reconciler/v1alpha1/service.(*Reconciler).reconcileConfiguration(0xc4208c64c0, 0x14f9880, 0xc4210ba0f0, 0xc4204e4540, 0xc4204b1400, 0x0, 0x50000c420184401, 0xc421080280)
        /tmp/upstream_src/src/github.com/knative/serving/pkg/reconciler/v1alpha1/service/service.go:241 +0x1af
github.com/knative/serving/pkg/reconciler/v1alpha1/service.(*Reconciler).reconcile(0xc4208c64c0, 0x14f9880, 0xc4210ba0f0, 0xc4204e4540, 0x0, 0x0)
        /tmp/upstream_src/src/github.com/knative/serving/pkg/reconciler/v1alpha1/service/service.go:166 +0x41d
github.com/knative/serving/pkg/reconciler/v1alpha1/service.(*Reconciler).Reconcile(0xc4208c64c0, 0x14f9880, 0xc4210ba0f0, 0xc420ff6620, 0x13, 0x694a3f59, 0x1d08420)
        /tmp/upstream_src/src/github.com/knative/serving/pkg/reconciler/v1alpha1/service/service.go:133 +0x653
github.com/knative/serving/vendor/github.com/knative/pkg/controller.(*Impl).processNextWorkItem.func1(0xc4208c6580, 0x119b360, 0xc420f86d30, 0x0, 0x0)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/knative/pkg/controller/controller.go:285 +0x328
github.com/knative/serving/vendor/github.com/knative/pkg/controller.(*Impl).processNextWorkItem(0xc4208c6580, 0x18)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/knative/pkg/controller/controller.go:295 +0x6a
github.com/knative/serving/vendor/github.com/knative/pkg/controller.(*Impl).Run.func1()
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/knative/pkg/controller/controller.go:219 +0x36
github.com/knative/serving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc420fee050)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/knative/serving/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc420fee050, 0x3b9aca00, 0x0, 0x1, 0xc42042c300)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
github.com/knative/serving/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc420fee050, 0x3b9aca00, 0xc42042c300)
        /tmp/upstream_src/src/github.com/knative/serving/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/knative/serving/vendor/github.com/knative/pkg/controller.(*Impl).Run
        /tmp/upstream_src/src/github.com/knative/serving/vendor/github.com/knative/pkg/controller/controller.go:218 +0x115
```

--- a/Gopkg.lock
+++ b/Gopkg.lock
@@ -396,7 +396,7 @@
   revision = "f0d7bb60956f88d4743f5fea66b5607b96d262c9"
 
 [[projects]]
-  digest = "1:74580f9e031cccd10526f3a47a24b10b61c358a690b5f1d531383c4963cf3311"
+  digest = "1:d53b75536c9a0fef9b6d8e7d152fa01841f413518db729216357b10f77b919e5"
   name = "github.com/knative/pkg"
   packages = [
     "apis",
@@ -426,6 +426,7 @@
     "configmap",
     "controller",
     "kmeta",
+    "kmp",
     "logging",
     "logging/logkey",
     "logging/testing",
@@ -440,7 +441,7 @@
     "webhook",
   ]
   pruneopts = "NUT"
-  revision = "0e41760cea1d1ce16f4cc311c2d6bd3525bb3da7"
+  revision = "3043da57e9f2d4f8afd664670357a09a80ebebdd"
 
 [[projects]]
   branch = "master"
@@ -1246,6 +1247,7 @@
     "github.com/knative/pkg/configmap",
     "github.com/knative/pkg/controller",
     "github.com/knative/pkg/kmeta",
+    "github.com/knative/pkg/kmp",
     "github.com/knative/pkg/logging",
     "github.com/knative/pkg/logging/logkey",
     "github.com/knative/pkg/logging/testing",

--- a/Gopkg.toml
+++ b/Gopkg.toml
@@ -23,8 +23,8 @@ required = [
 
 [[override]]
   name = "github.com/knative/pkg"
-  # HEAD as of 2018-12-05
-  revision = "0e41760cea1d1ce16f4cc311c2d6bd3525bb3da7"
+  # HEAD as of 2018-12-07
+  revision = "3043da57e9f2d4f8afd664670357a09a80ebebdd"
 
 [[override]]
   name = "go.uber.org/zap"

--- a/pkg/apis/autoscaling/v1alpha1/pa_validation.go
+++ b/pkg/apis/autoscaling/v1alpha1/pa_validation.go
@@ -19,8 +19,8 @@ package v1alpha1
 import (
 	"fmt"
 
-	"github.com/google/go-cmp/cmp"
 	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/kmp"
 	"github.com/knative/serving/pkg/apis/autoscaling"
 	servingv1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
 	autoscalingv1 "k8s.io/api/autoscaling/v1"
@@ -100,7 +100,13 @@ func (current *PodAutoscaler) CheckImmutableFields(og apis.Immutable) *apis.Fiel
 		return &apis.FieldError{Message: "The provided original was not a PodAutoscaler"}
 	}
 
-	if diff := cmp.Diff(original.Spec, current.Spec); diff != "" {
+	if diff, err := kmp.SafeDiff(original.Spec, current.Spec); err != nil {
+		return &apis.FieldError{
+			Message: "Failed to diff PodAutoscaler",
+			Paths:   []string{"spec"},
+			Details: err.Error(),
+		}
+	} else if diff != "" {
 		return &apis.FieldError{
 			Message: "Immutable fields changed (-old +new)",
 			Paths:   []string{"spec"},

--- a/pkg/apis/serving/v1alpha1/revision_validation.go
+++ b/pkg/apis/serving/v1alpha1/revision_validation.go
@@ -20,16 +20,14 @@ import (
 	"strconv"
 	"time"
 
-	"github.com/google/go-cmp/cmp"
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/kmp"
+	networkingv1alpha1 "github.com/knative/serving/pkg/apis/networking/v1alpha1"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/equality"
-	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/apimachinery/pkg/util/validation"
-
-	"github.com/knative/pkg/apis"
-	networkingv1alpha1 "github.com/knative/serving/pkg/apis/networking/v1alpha1"
 )
 
 func (rt *Revision) Validate() *apis.FieldError {
@@ -204,11 +202,13 @@ func (current *Revision) CheckImmutableFields(og apis.Immutable) *apis.FieldErro
 		return &apis.FieldError{Message: "The provided original was not a Revision"}
 	}
 
-	quantityComparer := cmp.Comparer(func(x, y resource.Quantity) bool {
-		return x.Cmp(y) == 0
-	})
-
-	if diff := cmp.Diff(original.Spec, current.Spec, quantityComparer); diff != "" {
+	if diff, err := kmp.SafeDiff(original.Spec, current.Spec); err != nil {
+		return &apis.FieldError{
+			Message: "Failed to diff Revision",
+			Paths:   []string{"spec"},
+			Details: err.Error(),
+		}
+	} else if diff != "" {
 		return &apis.FieldError{
 			Message: "Immutable fields changed (-old +new)",
 			Paths:   []string{"spec"},

--- a/pkg/reconciler/v1alpha1/revision/cruds.go
+++ b/pkg/reconciler/v1alpha1/revision/cruds.go
@@ -18,10 +18,10 @@ package revision
 
 import (
 	"context"
-
-	"github.com/google/go-cmp/cmp"
+	"fmt"
 
 	caching "github.com/knative/caching/pkg/apis/caching/v1alpha1"
+	"github.com/knative/pkg/kmp"
 	"github.com/knative/pkg/logging"
 	kpav1alpha1 "github.com/knative/serving/pkg/apis/autoscaling/v1alpha1"
 	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
@@ -83,8 +83,11 @@ func (c *Reconciler) checkAndUpdateService(ctx context.Context, rev *v1alpha1.Re
 	if equality.Semantic.DeepEqual(desiredService.Spec, service.Spec) {
 		return service, Unchanged, nil
 	}
-	logger.Infof("Reconciling service diff (-desired, +observed): %v",
-		cmp.Diff(desiredService.Spec, service.Spec))
+	diff, err := kmp.SafeDiff(desiredService.Spec, service.Spec)
+	if err != nil {
+		return nil, Unchanged, fmt.Errorf("failed to diff Service: %v", err)
+	}
+	logger.Infof("Reconciling service diff (-desired, +observed): %v", diff)
 
 	d, err := c.KubeClientSet.CoreV1().Services(service.Namespace).Update(desiredService)
 	return d, WasChanged, err

--- a/pkg/reconciler/v1alpha1/revision/reconcile_resources.go
+++ b/pkg/reconciler/v1alpha1/revision/reconcile_resources.go
@@ -21,7 +21,7 @@ import (
 	"fmt"
 	"time"
 
-	"github.com/google/go-cmp/cmp"
+	"github.com/knative/pkg/kmp"
 	"github.com/knative/pkg/logging"
 	"github.com/knative/pkg/logging/logkey"
 	kpav1alpha1 "github.com/knative/serving/pkg/apis/autoscaling/v1alpha1"
@@ -229,8 +229,11 @@ func (c *Reconciler) reconcileFluentdConfigMap(ctx context.Context, rev *v1alpha
 	} else {
 		desiredConfigMap := resources.MakeFluentdConfigMap(rev, cfgs.Observability)
 		if !equality.Semantic.DeepEqual(configMap.Data, desiredConfigMap.Data) {
-			logger.Infof("Reconciling fluentd configmap diff (-desired, +observed): %v",
-				cmp.Diff(desiredConfigMap.Data, configMap.Data))
+			diff, err := kmp.SafeDiff(desiredConfigMap.Data, configMap.Data)
+			if err != nil {
+				return fmt.Errorf("failed to diff ConfigMap: %v", err)
+			}
+			logger.Infof("Reconciling fluentd configmap diff (-desired, +observed): %v", diff)
 
 			// Don't modify the informers copy
 			existing := configMap.DeepCopy()

--- a/pkg/reconciler/v1alpha1/service/service.go
+++ b/pkg/reconciler/v1alpha1/service/service.go
@@ -18,10 +18,11 @@ package service
 
 import (
 	"context"
+	"fmt"
 	"reflect"
 
-	"github.com/google/go-cmp/cmp"
 	"github.com/knative/pkg/controller"
+	"github.com/knative/pkg/kmp"
 	"github.com/knative/pkg/logging"
 	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
 	servinginformers "github.com/knative/serving/pkg/client/informers/externalversions/serving/v1alpha1"
@@ -248,7 +249,11 @@ func (c *Reconciler) reconcileConfiguration(ctx context.Context, service *v1alph
 		// No differences to reconcile.
 		return config, nil
 	}
-	logger.Infof("Reconciling configuration diff (-desired, +observed): %v", cmp.Diff(desiredConfig.Spec, config.Spec))
+	diff, err := kmp.SafeDiff(desiredConfig.Spec, config.Spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to diff Configuration: %v", err)
+	}
+	logger.Infof("Reconciling configuration diff (-desired, +observed): %v", diff)
 
 	// Don't modify the informers copy.
 	existing := config.DeepCopy()
@@ -285,7 +290,11 @@ func (c *Reconciler) reconcileRoute(ctx context.Context, service *v1alpha1.Servi
 		// No differences to reconcile.
 		return route, nil
 	}
-	logger.Infof("Reconciling route diff (-desired, +observed): %v", cmp.Diff(desiredRoute.Spec, route.Spec))
+	diff, err := kmp.SafeDiff(desiredRoute.Spec, route.Spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to diff Route: %v", err)
+	}
+	logger.Infof("Reconciling route diff (-desired, +observed): %v", diff)
 
 	// Don't modify the informers copy.
 	existing := route.DeepCopy()

